alacritty_terminal/src/term/text_run.rs

```diff
@@ -0,0 +1,309 @@
+use super::{
+    cell::{Flags, MAX_ZEROWIDTH_CHARS},
+    color::Rgb,
+    CursorKey, Point, RenderableCell, RenderableCellContent,
+};
+use crate::index::{Column, Line};
+
+#[derive(Debug)]
+struct RunStart {
+    pub line: Line,
+    pub column: Column,
+    pub fg: Rgb,
+    pub bg: Rgb,
+    pub bg_alpha: f32,
+    pub flags: Flags,
+}
+
+// Use a macro instead of a function to make use of partial move semantics that don't cross
+// function boundaries
+// Convert a RenderableCell into a RunStart
+macro_rules! from_rc {
+    ($rc:ident) => {
+        RunStart {
+            line: $rc.line,
+            column: $rc.column,
+            fg: $rc.fg,
+            bg: $rc.bg,
+            bg_alpha: $rc.bg_alpha,
+            flags: $rc.flags,
+        }
+    };
+}
+
+/// Compare cells and check they are in the same text run
+fn is_adjacent_context(a: &RunStart, b: &RenderableCell) -> bool {
+    a.line == b.line
+        && a.fg == b.fg
+        && a.bg == b.bg
+        && (a.bg_alpha - b.bg_alpha).abs() < std::f32::EPSILON
+        && a.flags == b.flags
+}
+
+type LatestCol = (Column, bool);
+/// Checks two columns are adjacent
+fn is_adjacent_col((a, is_wide): LatestCol, b: Column) -> bool {
+    let width = if is_wide { 2usize } else { 1usize };
+    a + width == b || b + width == a
+}
+
+#[derive(Debug)]
+pub enum TextRunContent {
+    Cursor(CursorKey),
+    CharRun(String, Vec<[char; MAX_ZEROWIDTH_CHARS]>),
+}
+
+/// Represents a set of renderable cells that all share the same rendering propreties.
+/// The assumption is that if two cells are in the same TextRun they can be sent off together to
+/// be shaped. This allows for ligatures to be rendered but not when something breaks up a ligature
+/// (e.g. selection hightlight) which is desired behavior.
+#[derive(Debug)]
+pub struct TextRun {
+    /// By definition a run is on one line.
+    pub line: Line,
+    /// Span of columns the text run covers.
+    pub span: (Column, Column),
```

Comment:
I don't see why this wouldn't be possible?

```diff
diff --git a/alacritty_terminal/src/renderer/mod.rs b/alacritty_terminal/src/renderer/mod.rs
index fff5187..c3e5243 100644
--- a/alacritty_terminal/src/renderer/mod.rs
+++ b/alacritty_terminal/src/renderer/mod.rs
@@ -1042,7 +1042,7 @@ impl<'a> RenderApi<'a> {
 
         let text_run = TextRun {
             line,
-            span: (Column(0), Column(string.len() - 1)),
+            span: Column(0)..=Column(string.len() - 1),
             content: TextRunContent::CharRun(string.to_owned(), vec![]),
             fg: Rgb { r: 0, g: 0, b: 0 },
             bg: color.unwrap_or(Rgb { r: 0, g: 0, b: 0 }),
diff --git a/alacritty_terminal/src/text_run.rs b/alacritty_terminal/src/text_run.rs
index 4099647..c38dd6c 100644
--- a/alacritty_terminal/src/text_run.rs
+++ b/alacritty_terminal/src/text_run.rs
@@ -41,7 +41,7 @@ pub struct TextRun {
     /// A run never spans multiple lines.
     pub line: Line,
     /// Span of columns the text run covers.
-    pub span: (Column, Column),
+    pub span: std::ops::RangeInclusive<Column>,
     /// Cursor or sequence of characters.
     pub content: TextRunContent,
     /// Foreground color of text run content.
@@ -63,7 +63,7 @@ impl TextRun {
         let end_column = if is_wide { latest + 1 } else { latest };
         TextRun {
             line: start.line,
-            span: (start.column, end_column),
+            span: start.column..=end_column,
             content: TextRunContent::CharRun(buffer.0, buffer.1),
             fg: start.fg,
             bg: start.bg,
@@ -75,7 +75,7 @@ impl TextRun {
     fn from_cursor_rc(start: RunStart, cursor: CursorKey) -> Self {
         TextRun {
             line: start.line,
-            span: (start.column, start.column),
+            span: start.column..=start.column,
             content: TextRunContent::Cursor(cursor),
             fg: start.fg,
             bg: start.bg,
@@ -99,24 +99,25 @@ impl TextRun {
 
     /// First cell in the TextRun
     pub fn start_cell(&self) -> RenderableCell {
-        self.cell_at(self.span.0)
+        self.cell_at(*self.span.start())
     }
 
     /// First point covered by this TextRun
     pub fn start_point(&self) -> Point {
-        Point { line: self.line, col: self.span.0 }
+        Point { line: self.line, col: *self.span.start() }
     }
 
     /// End point covered by this TextRun
     pub fn end_point(&self) -> Point {
-        Point { line: self.line, col: self.span.1 }
+        Point { line: self.line, col: *self.span.end() }
     }
 
     /// Iterates over each RenderableCell in column range [run.0, run.1]
     pub fn cells<'a>(&'a self) -> impl Iterator<Item = RenderableCell> + 'a {
         let step = if self.flags.contains(Flags::WIDE_CHAR) { 2 } else { 1 };
-        let (Column(start), Column(end)) = self.span;
+
         // TODO: impl Step for Column (once Step is stable) to avoid unwrapping then rewrapping.
+        let (&Column(start), &Column(end)) = (self.span.start(), self.span.end());
         (start..=end).step_by(step).map(move |col| self.cell_at(Column(col)))
     }
 }
```